### PR TITLE
[ie/GodResource] Add extractor

### LINKED ISSUE #9551: God Resource
### DO NOT REMOVE OR SKIP THE ISSUE TEMPLATE

- [X] I understand that I will be **blocked** if I *intentionally* remove or skip any mandatory\* field

### Checklist

- [X] I'm reporting a new site support request
- [X] I've verified that I have **updated yt-dlp to nightly or master** ([update instructions](https://github.com/yt-dlp/yt-dlp#update-channels))
- [X] I've checked that all provided URLs are playable in a browser with the same IP and same login details
- [X] I've checked that none of provided URLs [violate any copyrights](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy) or contain any [DRM](https://en.wikipedia.org/wiki/Digital_rights_management) to the best of my knowledge
- [X] I've searched [known issues](https://github.com/yt-dlp/yt-dlp/issues/3766) and the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues **including closed ones**. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and am willing to share it if required

### Region

United States 

### Example URLs

single video - https://new.godresource.com/video/A01mTKjyf6w

### Provide a description that is worded well enough to be understood

Please provide support for the website,  https://www.godresource.com/ and https://new.godresource.com.  Thanks.

### Provide verbose output that clearly demonstrates the problem

- [X] Run **your** yt-dlp command with **-vU** flag added (`yt-dlp -vU <your command line>`)
- [ ] If using API, add `'verbose': True` to `YoutubeDL` params instead
- [X] Copy the WHOLE output (starting with `[debug] Command-line config`) and insert it below

### Complete Verbose Output

```shell
[debug] Command-line config: ['-vU', 'https://new.godresource.com/video/A01mTKjyf6w']
[debug] Encodings: locale utf-8, fs utf-8, pref utf-8, out utf-8, error utf-8, screen utf-8
[debug] yt-dlp version stable@2024.03.10 from yt-dlp/yt-dlp [615a84447] (pip)
[debug] Python 3.11.8 (CPython aarch64 64bit) - Linux-4.19.157-perf+-aarch64-with-libc (OpenSSL 3.2.1 30 Jan 2024, libc)
[debug] exe versions: ffmpeg 6.1.1 (setts), ffprobe 6.1.1
[debug] Optional libraries: Cryptodome-3.19.0, brotli-1.1.0, certifi-2023.07.22, mutagen-1.47.0, requests-2.31.0, sqlite3-3.45.2, urllib3-2.1.0, websockets-12.0
[debug] Proxy map: {}
[debug] Request Handlers: urllib, requests, websockets
[debug] Loaded 1803 extractors
[debug] Fetching release info: https://api.github.com/repos/yt-dlp/yt-dlp/releases/latest
Latest version: stable@2024.03.10 from yt-dlp/yt-dlp
yt-dlp is up to date (stable@2024.03.10 from yt-dlp/yt-dlp)
[generic] Extracting URL: https://new.godresource.com/video/A01mTKjyf6w
[generic] A01mTKjyf6w: Downloading webpage
ERROR: [generic] Unable to download webpage: HTTP Error 404: Not Found (caused by <HTTPError 404: Not Found>)
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/yt_dlp/extractor/common.py", line 732, in extract
    ie_result = self._real_extract(url)
                ^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/yt_dlp/extractor/generic.py", line 2392, in _real_extract
    full_response = self._request_webpage(url, video_id, headers=filter_dict({
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/yt_dlp/extractor/common.py", line 877, in _request_webpage
    raise ExtractorError(errmsg, cause=err)

  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/yt_dlp/extractor/common.py", line 864, in _request_webpage
    return self._downloader.urlopen(self._create_request(url_or_request, data, headers, query))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/yt_dlp/YoutubeDL.py", line 4101, in urlopen
    return self._request_director.send(req)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/yt_dlp/networking/common.py", line 115, in send
    response = handler.send(request)
               ^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/yt_dlp/networking/_helper.py", line 204, in wrapper
    return func(self, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/yt_dlp/networking/common.py", line 326, in send
    return self._send(request)
           ^^^^^^^^^^^^^^^^^^^
  File "/data/data/com.termux/files/usr/lib/python3.11/site-packages/yt_dlp/networking/_requests.py", line 351, in _send
    raise HTTPError(res, redirect_loop=max_redirects_exceeded)
yt_dlp.networking.exceptions.HTTPError: HTTP Error 404: Not Found
```

### PR DESCRIPTION
**Pull request by HobbyistDev**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

This PR add GodResource extractor. This extractor only extract in `news.godresource.com` because i didn't have any link in `www.godresource.com` that have the video. This website is not giving reliable `title` so it just grab the `title` from html `title` tag (which is default to the site name, `God Resource`.

Fixes #9551

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check all of the following options that apply:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [ ] Fix or improvement to an extractor (Make sure to add/update tests)
- [x] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

</details>

### CODE REVIEW

File: yt_dlp/extractor/godresource.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+
+            extraction_result = {
+                'formats': formats,
+                'subtitles': subtitles
+            }
+        elif determine_ext(video_url) == 'mp4':
+            extraction_result = {
+                'url': video_url,
+                'ext': 'mp4'
+            }
```
Comment (pukkandan) [inline yt_dlp/extractor/godresource.py:68]:
```suggestion
            extraction_result = {'formats': [{
                'url': video_url,
                'ext': 'mp4'
            }]}
```
is prefered

File: yt_dlp/extractor/godresource.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+            'title': 'The Sodomite Deception',
+            'upload_date': '20230629',
+        }
+    }]
+
+    def _real_extract(self, url):
+        display_id = self._match_id(url)
+
+        # the website is oddly giving all request as 404 at first and then loaded with js
+        webpage, _ = self._download_webpage_handle(url, display_id, expected_status=404)
```
Comment (pukkandan) [inline yt_dlp/extractor/godresource.py:47]:
```suggestion
        # the website is oddly giving all request as 404 at first and then loaded with js
        webpage = self._download_webpage(url, display_id, expected_status=404)
```

Comment (pukkandan) [reply yt_dlp/extractor/godresource.py:47]: This is only used for the fallback title. Do we really need it? Imo,we can just skip this download completely

```suggestion
```

File: yt_dlp/extractor/godresource.py (hunk 1 above)
Comment (pukkandan) [inline yt_dlp/extractor/godresource.py:68]:
```suggestion
        video_url = api_data['streamUrl']
        if (ext := determine_ext(video_url)) == 'm3u8':
            formats, subtitles = self._extract_m3u8_formats_and_subtitles(
                api_data['streamUrl'], display_id)
        elif ext == 'mp4':
            formats, subtitles = [{
                'url': video_url,
                'ext': ext
            }], {}
        else:
            raise ExtractorError(f'Unexpected video format {ext}', True)
```

File: yt_dlp/extractor/godresource.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+            }
+        elif determine_ext(video_url) == 'mp4':
+            extraction_result = {
+                'url': video_url,
+                'ext': 'mp4'
+            }
+
+        return {
+            'id': display_id,
+            **extraction_result,
```
Comment (pukkandan) [inline yt_dlp/extractor/godresource.py:72]:
```suggestion
            'formats': formats,
            'subtitles': subtitles,
```

File: yt_dlp/extractor/godresource.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                'thumbnail': ('thumbnail', {url_or_none}),
+                'view_count': ('views', {int}),
+                'channel': ('channelName', {str}),
+                'channel_id': 'channelId',
+                'timestamp': ('streamDateCreated', {unified_timestamp}),
+                'modified_timestamp': ('streamDataModified', {unified_timestamp})
+            }), {
+                # This title is not reliable (only give the site name)
+                'title': self._html_extract_title(webpage),
+            }),
```
Comment (pukkandan) [inline yt_dlp/extractor/godresource.py:84]: traverse_obj doesnt actually return the None. So merge_dicts is unncessary

```suggestion
            # This title is not reliable (only give the site name)
            'title': self._html_extract_title(webpage),
            **traverse_obj(api_data, {
                'title': ('title', {str}),  # can be None in API
                'thumbnail': ('thumbnail', {url_or_none}),
                'view_count': ('views', {int}),
                'channel': ('channelName', {str}),
                'channel_id': ('channelId', {str_or_none}),
                'timestamp': ('streamDateCreated', {unified_timestamp}),
                'modified_timestamp': ('streamDataModified', {unified_timestamp})
            })
```

File: yt_dlp/extractor/godresource.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+        'info_dict': {
+            'id': 'A01mTKjyf6w',
+            'ext': 'mp4',
+            'view_count': int,
+            'timestamp': 1710978666,
+            'channel_id': 5,
+            'thumbnail': 'https://cdn-02.godresource.com/e42968ac-9e8b-4231-ab86-f4f9d775841f/thumbnail.jpg',
+            'channel': 'Stedfast Baptist Church',
+            'upload_date': '20240320',
+            'title': 'GodResource',
```
Comment (pukkandan) [inline yt_dlp/extractor/godresource.py:25]: This is worse than the auto-generated title 'GodResource video #A01mTKjyf6w' and should not be used imo

File: yt_dlp/extractor/godresource.py (hunk 4 above)
Comment (pukkandan) [reply yt_dlp/extractor/godresource.py:84]: If skipping fallback title, you can add `'title': ''` to suppress warning

```suggestion
            'title': '',
            **traverse_obj(api_data, {
                'title': ('title', {str}),  # can be None in API
                'thumbnail': ('thumbnail', {url_or_none}),
                'view_count': ('views', {int}),
                'channel': ('channelName', {str}),
                'channel_id': ('channelId', {str_or_none}),
                'timestamp': ('streamDateCreated', {unified_timestamp}),
                'modified_timestamp': ('streamDataModified', {unified_timestamp})
            })
```

File: yt_dlp/extractor/godresource.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+    _VALID_URL = r'https?://new\.godresource\.com/video/(?P<id>\w+)'
+    _TESTS = [{
+        # hls stream
+        'url': 'https://new.godresource.com/video/A01mTKjyf6w',
+        'info_dict': {
+            'id': 'A01mTKjyf6w',
+            'ext': 'mp4',
+            'view_count': int,
+            'timestamp': 1710978666,
+            'channel_id': 5,
```
Comment (pukkandan) [inline yt_dlp/extractor/godresource.py:21]: Should be string

**Review by pukkandan — CHANGES_REQUESTED** [review]

File: yt_dlp/extractor/godresource.py (hunk 1 above)
Comment (pukkandan) [reply yt_dlp/extractor/godresource.py:68]: superseeded by other comment

File: yt_dlp/extractor/godresource.py (hunk 3 above)
Comment (HobbyistDev) [reply yt_dlp/extractor/godresource.py]: Fixed in f326e05e9f027466a9914a9724364979c3321256

File: yt_dlp/extractor/godresource.py (hunk 5 above)
Comment (HobbyistDev) [reply yt_dlp/extractor/godresource.py]: Resolved in d493ce72b63483e20d8001f8eabcc286cd4eb3b2

File: yt_dlp/extractor/godresource.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+            'channel_id': '5',
+            'thumbnail': 'https://cdn-02.godresource.com/e42968ac-9e8b-4231-ab86-f4f9d775841f/thumbnail.jpg',
+            'channel': 'Stedfast Baptist Church',
+            'upload_date': '20240320',
+            'title': 'GodResource video #A01mTKjyf6w',
+        }
+    }, {
+        # mp4 link
+        'url': 'https://new.godresource.com/video/01DXmBbQv_X',
+        'info_dict': {
```
Comment (pukkandan) [inline yt_dlp/extractor/godresource.py:31]: An md5 check can be added for this (they don't work for hls)
```diff
        'url': 'https://new.godresource.com/video/01DXmBbQv_X',
+       'md5': 'xxx',
```

**Review by pukkandan — APPROVED** [review]

File: yt_dlp/extractor/godresource.py (hunk 7 above)
Comment (HobbyistDev) [reply yt_dlp/extractor/godresource.py:32]: Fixed in 4741aded6fda773eefc3b6d53335ea26b9d36b4a

File: yt_dlp/extractor/godresource.py [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+        }
+    }]
+
+    def _real_extract(self, url):
+        display_id = self._match_id(url)
+
+        api_data = self._download_json(
+            f'https://api.godresource.com/api/Streams/{display_id}', display_id)
+
+        video_url = api_data['streamUrl']
```
Comment (seproDev) [inline yt_dlp/extractor/godresource.py:51]:
```suggestion
        video_url = api_data['streamUrl']
        is_live = api_data.get('isLive') or False
```

File: yt_dlp/extractor/godresource.py [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+        if (ext := determine_ext(video_url)) == 'm3u8':
+            formats, subtitles = self._extract_m3u8_formats_and_subtitles(
+                api_data['streamUrl'], display_id)
```
Comment (seproDev) [inline yt_dlp/extractor/godresource.py:54]:
```suggestion
                video_url, display_id, live=is_live)
```

File: yt_dlp/extractor/godresource.py [hunk 10]
```diff
@@ -0,0 +1,77 @@
...
+                'ext': ext
+            }], {}
+        else:
+            raise ExtractorError(f'Unexpected video format {ext}')
+
+        return {
+            'id': display_id,
+            'formats': formats,
+            'subtitles': subtitles,
+            'title': '',
```
Comment (seproDev) [inline yt_dlp/extractor/godresource.py:67]:
```suggestion
            'title': '',
            'is_live': is_live,
```

**Review by seproDev — CHANGES_REQUESTED** [review]
Otherwise, ongoing livestreams get downloaded as regular videos.

**Review by Grub4K — APPROVED** [review]

**Review by bashonly — APPROVED** [review]

**Pull request merged**